src/daft-io/src/object_io.rs

```diff
@@ -112,10 +118,240 @@ pub(crate) trait ObjectSource: Sync + Send {
+                // Last fragment does not contain wildcard: we return it if the full path exists and is a FileType::File
+                } else {
+                    let full_dir_path = state.current_path.clone() + current_fragment.escaped_str();
+                    let single_file_ls = source.ls(full_dir_path.as_str(), Some("/"), None).await;
+                    match single_file_ls {
+                        Ok(mut single_file_ls) => {
+                            if single_file_ls.files.len() == 1
+                                && matches!(single_file_ls.files[0].filetype, FileType::File)
+                            {
+                                let fm = single_file_ls.files.drain(..).next().unwrap();
+                                result_tx.send(Ok(fm)).await.expect("Internal multithreading channel is broken: results may be incorrect");
+                            }
+                        }
+                        Err(e) => {
+                            // Silence NotFound errors when in wildcard "search" mode
+                            if !(matches!(e, super::Error::NotFound { .. }) && state.wildcard_mode)
+                            {
+                                result_tx.send(Err(e)).await.expect(
+                                    "Internal multithreading channel is broken: results may be incorrect",
+                                );
+                            }
+                        }
+                    };
+                }
+
+            // RECURSIVE CASE: current_fragment contains a special character (e.g. *)
+            } else if current_fragment.has_special_character() {
+                let partial_glob_matcher = GlobBuilder::new(
+                    GlobFragment::join(
+                        &state.glob_fragments[..state.current_fragment_idx + 1],
+                        "/",
+                    )
+                    .raw_str(),
+                )
+                .literal_separator(true)
+                .build()
+                .expect("Cannot parse glob")
+                .compile_matcher();
+                let mut results = source
+                    .iter_dir(&state.current_path, Some("/"), None)
+                    .await
+                    .unwrap_or_else(|e| stream! {yield Err(e)}.boxed());
+
+                while let Some(val) = results.next().await {
+                    match val {
+                        Ok(fm) => {
+                            if matches!(fm.filetype, FileType::Directory)
+                                && partial_glob_matcher
+                                    .is_match(fm.filepath.as_str().trim_end_matches('/'))
+                            {
+                                visit(
+                                    result_tx.clone(),
+                                    source.clone(),
+                                    state
+                                        .advance(fm.filepath, state.current_fragment_idx + 1)
```

Comment:
this is a double clone. converting to consuming functions would fix that

Reply:
`.advance` can't consume: it has to be called multiple times because the current state is advanced multiple times, one for each `visit` call called to a subdirectory/path

I'll switch `with_wildcard_mode` into a consuming call though to avoid a clone there

Reply:
advance should consume or take a &mut self, and instead of an implicit clone, it should be explicit. Think of the case that you already have ownership of the state, it would be weird to clone something you already own. And in the case of a loop, it makes sense to clone explicitly since the first loop moves the value.

Reply:
Check out this article for the design pattern: https://hoverbear.org/blog/rust-state-machine-pattern/